File: typescript/src/agent.ts
/**
 * Copyright 2026 ProvnAI
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as crypto from 'crypto';
const { canonicalize } = require('json-canonicalize');
import axios from 'axios';
import { VEPBuilder, VexCapsule, IntentSegment, AuthoritySegment } from './builder';
import { ProvnSDK, init } from '@provncloud/sdk';
const { CipherSuite } = require('hpke-js');
const { DhkemX25519HkdfSha256, HkdfSha256 } = require('@hpke/dhkem-x25519');
const { Aes128Gcm } = require('@hpke/core');

export interface VexConfig {
    identityKey: string;     // Hex or Base64 Ed25519 private key
    vanguardUrl: string;     // Target proxy endpoint (McpVanguard/VEX sidecar)
    aid?: string;            // Agent Identity ID (NEW in v3)
}

export interface VexResult {
    status: string;
    outcome: 'ALLOW' | 'HALT' | 'ESCALATE';
    reason_code?: string;
    capsule_root: string;
    capability_token?: string;
    [key: string]: unknown;
}

export class VexAgent {
    private config: VexConfig;
    private sdk: ProvnSDK | null = null;

    private currentToken?: string;

    constructor(config: VexConfig) {
        this.config = config;
    }

    private async ensureSDK() {
        if (!this.sdk) {
            await init();
            this.sdk = new ProvnSDK();
        }
    }

    /**
     * Retrieves the Gate's public key for HPKE encryption.
     */
    async fetchPublicKey(): Promise<string> {
        const response = await axios.get(`${this.config.vanguardUrl}/public_key`);
        return response.data.public_key;
    }

    /**
     * Executes a tool via the VEX verifiable execution loop.
     */
    /**
     * Locally verifies a VEX Continuation Token (v3) against the Gate's public key.
     * Ensures the token was signed by the authoritative Gate and binds to the current capsule.
     */
    async verifyToken(tokenBase64: string, expectedCapsuleRoot?: string): Promise<boolean> {
        const { canonicalize } = require('json-canonicalize');
        const crypto = require('crypto');
        
        try {
            const token = JSON.parse(Buffer.from(tokenBase64, 'base64').toString());
            const gatePkBase64 = await this.fetchPublicKey();
            const gatePublicKey = crypto.createPublicKey({
                key: Buffer.from(gatePkBase64, 'base64'),
                format: 'der',
                type: 'spki', // Ed25519 in SPKI format
            });

            // 1. Re-hash the payload (JCS)
            const payloadHash = crypto.createHash('sha256')
                .update(canonicalize(token.payload))
                .digest();

            // 2. Verify signature
            const isSignatureValid = crypto.verify(
                null,
                payloadHash,
                gatePublicKey,
                Buffer.from(token.signature, 'hex')
            );

            if (!isSignatureValid) {
                console.error('VEX Token Verification Failed: Invalid Signature');
                return false;
            }

            // 3. Bind to capsule root
            if (expectedCapsuleRoot && token.payload.source_capsule_root !== expectedCapsuleRoot) {
                console.error('VEX Token Verification Failed: Root Mismatch');
                return false;
            }

            return true;
        } catch (error) {
            console.error('VEX Token Verification Error:', error);
            return false;
        }
    }

    async execute(toolName: string, params: Record<string, unknown>, intentContext?: string): Promise<VexResult> {
        await this.ensureSDK();

        // 1. Build Capsule
        const capsuleData = await this.buildCapsule(toolName, params, intentContext);
        const capsuleId = capsuleData.authority.capsule_id;

        // 2. Dispatch to Vanguard
        try {
            const response = await axios.post(`${this.config.vanguardUrl}/dispatch`, capsuleData, {
                headers: { 'Content-Type': 'application/json' }
            });
            let result = response.data;

            // 3. Handle AEM (ESCALATE Loop)
            let attempts = 0;
            const maxAttempts = 15; // ~30 seconds

            while (result.outcome === 'ESCALATE' && attempts < maxAttempts) {
                await new Promise(resolve => setTimeout(resolve, 2000));
                attempts++;

                const pollResp = await axios.get(`${this.config.vanguardUrl}/capsule/${capsuleId}/status`, {
                    headers: { 'Accept': 'application/json' }
                });
                result = pollResp.data;
            }

            if (result.outcome === 'HALT') {
                throw new Error(`VEX Execution HALTED: ${result.reason_code}`);
            }

            if (result.outcome === 'ESCALATE') {
                throw new Error('VEX Escalated resolution timed out.');
            }

            // 4. Store Capability Token
            if (result.capability_token) {
                this.currentToken = result.capability_token;
            }

            return result as VexResult;
        } catch (error) {
            const err = error as { response?: { data: Record<string, unknown> }, message: string };
            console.error('VEX Execution Failed:', err.response?.data || err.message);
            throw error;
        }
    }

    /**
     * Manually construct a signed Evidence Capsule without dispatching it.
     */
    async buildCapsule(toolName: string, params: Record<string, unknown>, intentContext?: string): Promise<VexCapsule> {
        await this.ensureSDK();
        // This is a simplified implementation for the SDK.
        // It demonstrates how the builder is leveraged.
        
        const intent: IntentSegment = {
            schema: 'vex/intent/v3',
            aid: this.config.aid || '00'.repeat(32),
            request_sha256: this.hashObject(params),
            commands: [toolName, params],
            confidence: 1.0,
            capabilities: ['sdk_execution']
        };
        if (intentContext) {
            intent.intent_context = intentContext;
        }

        const authority: AuthoritySegment = {
            capsule_id: require('crypto').randomUUID(),
            outcome: 'ALLOW',
            reason_code: 'SDK_GENERATED',
            trace_root: '00'.repeat(32),
            nonce: Date.now(),
            prev_hash: '00'.repeat(32), // Start of chain
            binding_status: 'UNBOUND',
            supervision: {
                branch_completeness: 1.0,
                contradictions: 0,
                confidence: 1.0
            }
        };

        const identity = {
            aid: this.config.aid || '00'.repeat(32),
            identity_type: 'software_sim',
            pcrs: { '0': '00'.repeat(32) }
        };

        const witness = {
            chora_node_id: 'local_witness',
            receipt_hash: '00'.repeat(32),
            timestamp: Math.floor(Date.now() / 1000)
        };

        // --- Phase 2: HPKE Encryption (Optional/v3) ---
        let intent_hash: string;
        const gatePkBase64 = await this.fetchPublicKey().catch(() => null);
        
        if (gatePkBase64) {
            const suite = new CipherSuite({
                kem: new DhkemX25519HkdfSha256(),
                kdf: new HkdfSha256(),
                aead: new Aes128Gcm(),
            });

            const recipientPublicKeyRaw = new Uint8Array(Buffer.from(gatePkBase64, 'base64'));
            const recipientKey = await suite.importKey('raw', recipientPublicKeyRaw, true);
            const info = new Uint8Array(Buffer.from('vex/intent/v3'));
            
            const { enc, ct } = await suite.seal(
                { 
                    recipientPublicKey: recipientKey,
                    info: info
                },
                new Uint8Array(Buffer.from(JSON.stringify(intent)))
            );

            // In v1.6.0, the "Intent Pillar" commitment is the hash of the ciphertext
            intent_hash = require('crypto').createHash('sha256').update(Buffer.from(ct)).digest('hex');
            
            // Add HPKE metadata to the segment so the Gate can decrypt
            intent.hpke = {
                enc: Buffer.from(enc).toString('base64'),
                ciphertext: Buffer.from(ct).toString('base64'),
                schema: 'vex/intent/v3/encrypted'
            };
        } else {
            // Fallback: Standard JCS hash for Transparent intents
            intent_hash = VEPBuilder.hashSegment(intent);
        }

        const authority_hash = VEPBuilder.hashSegment(authority);
        const identity_hash = VEPBuilder.hashSegment(identity);
        const witness_hash = VEPBuilder.hashSegment(witness, false); // Minimal scope

        const capsule_root = VEPBuilder.calculateCapsuleRoot({
            intent_hash,
            authority_hash,
            identity_hash,
            witness_hash
        });

        // Sign the capsule_root
        const claim = this.sdk!.createClaimNow(capsule_root);
        const keyPair = { 
            private_key: this.config.identityKey, 
            public_key: '00'.repeat(32) // In real use, this comes from the config/key derivation
        };
        this.sdk!.setKeypair(keyPair);
        const signedClaim = this.sdk!.signClaim(claim);
        const signature = Buffer.from(signedClaim.signature, 'hex');

        return {
            intent,
            authority,
            identity,
            witness,
            intent_hash,
            authority_hash,
            identity_hash,
            witness_hash,
            capsule_root,
            crypto: {
                algo: 'ed25519',
                signature_scope: 'capsule_root',
                signature_b64: signature.toString('base64'),
                signature_raw: signature // Keep raw for binary spec
            }
        };
    }

    /**
     * Serializes the Evidence Capsule into the v0x03 Binary Wire format.
     */
    toBinary(capsule: VexCapsule): Buffer {
        const { canonicalize } = require('json-canonicalize');
        
        // --- Header (76 Bytes) ---
        const header = Buffer.alloc(76);
        header.write('VEP', 0);
        header.writeUInt8(0x03, 3);
        header.write(capsule.identity.aid.replace(/-/g, ''), 4, 'hex'); // Ensure hex
        header.write(capsule.capsule_root, 36, 'hex');
        header.writeBigUInt64BE(BigInt(capsule.authority.nonce), 68);

        // --- TLV Body ---
        const packTLV = (tag: number, data: Buffer) => {
            const tlv = Buffer.alloc(5 + data.length);
            tlv.writeUInt8(tag, 0);
            tlv.writeUInt32BE(data.length, 1);
            data.copy(tlv, 5);
            return tlv;
        };

        const segments = [
            packTLV(0x01, Buffer.from(canonicalize(capsule.intent))),
            packTLV(0x02, Buffer.from(canonicalize(capsule.authority))),
            packTLV(0x03, Buffer.from(canonicalize(capsule.identity))),
            packTLV(0x05, Buffer.from(canonicalize(capsule.witness))),
            packTLV(0x06, capsule.crypto.signature_raw)
        ];

        return Buffer.concat([header, ...segments, header]);
    }

    private hashObject(obj: Record<string, unknown>): string {
        const canonicalJSON = canonicalize(obj);
        return crypto.createHash('sha256').update(canonicalJSON).digest('hex');
    }
}

File: python/provn_vex_sdk/agent.py
# Copyright 2026 ProvnAI
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import httpx
import uuid
import time
import provn_sdk
import base64
import json
import hashlib
from typing import Dict, Any, Optional
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import x25519, ed25519
from cryptography.hazmat.primitives.kdf.hkdf import HKDF
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from provn_vex_sdk.builder import VEPBuilder, IntentSegment, AuthoritySegment, IdentitySegment, WitnessSegment

class VexAgent:
    def __init__(self, identity_key: str, vanguard_url: str = "http://localhost:3000", aid: Optional[str] = None):
        self.identity_key = identity_key
        self.vanguard_url = vanguard_url
        self.aid = aid or "00" * 32
        self.current_token = None
        self._gate_pk_cache = None

    async def fetch_public_key(self) -> str:
        """Retrieves the Gate's public key for HPKE encryption."""
        async with httpx.AsyncClient() as client:
            response = await client.get(f"{self.vanguard_url}/public_key")
            response.raise_for_status()
            return response.json()["public_key"]

    async def verify_token(self, token_b64: str, expected_capsule_root: Optional[str] = None) -> bool:
        """Locally verifies a VEX Continuation Token (v3) against the Gate's public key."""
        import jcs
        try:
            token_json = base64.b64decode(token_b64).decode()
            token = json.loads(token_json)
            
            gate_pk_b64 = await self.fetch_public_key()
            gate_pk_bytes = base64.b64decode(gate_pk_b64)
            
            # Ed25519 Public Key from SPKI/Raw bytes
            # VEX uses raw 32-byte Ed25519 keys or SPKI. Assuming raw for now based on VEX v1.6.0.
            try:
                public_key = ed25519.Ed25519PublicKey.from_public_bytes(gate_pk_bytes)
            except Exception:
                # Fallback to SPKI
                public_key = serialization.load_der_public_key(gate_pk_bytes)

            # 1. Re-hash payload
            payload_jcs = jcs.canonicalize(token["payload"])
            payload_hash = hashlib.sha256(payload_jcs).digest()

            # 2. Verify signature
            signature = bytes.fromhex(token["signature"])
            public_key.verify(signature, payload_hash)

            # 3. Bind to capsule root
            if expected_capsule_root and token["payload"]["source_capsule_root"] != expected_capsule_root:
                return False

            return True
        except Exception as e:
            print(f"VEX Token Verification Failed: {str(e)}")
            return False

    async def execute(self, tool_name: str, parameters: Dict[str, Any], intent_context: Optional[str] = None) -> Dict[str, Any]:
        """
        Executes a tool via the VEX verifiable execution loop.
        Includes AEM (Authorization Enforcement Module) handshake and ESCALATE handling.
        """
        import asyncio
        
        # 1. Build Capsule
        capsule_data = await self.build_capsule(tool_name, parameters, intent_context)
        capsule_id = capsule_data["authority"]["capsule_id"]

        # 2. Dispatch to Vanguard
        async with httpx.AsyncClient() as client:
            try:
                response = await client.post(
                    f"{self.vanguard_url}/dispatch",
                    json=capsule_data,
                    headers={"Content-Type": "application/json"}
                )
                response.raise_for_status()
                result = response.json()

                # 3. Handle AEM (ESCALATE Loop)
                attempts = 0
                max_attempts = 15 
                backoff = 1.0
                
                while result.get("outcome") == "ESCALATE" and attempts < max_attempts:
                    await asyncio.sleep(backoff)
                    attempts += 1
                    backoff = min(backoff * 1.5, 5.0) # Exponential backoff capped at 5s
                    
                    # Poll for Resolution Capsule
                    try:
                        resp = await client.get(
                            f"{self.vanguard_url}/capsule/{capsule_id}/status",
                            timeout=10.0
                        )
                        resp.raise_for_status()
                        result = resp.json()
                    except httpx.HTTPError as e:
                        print(f"[VEX] Polling error (attempt {attempts}): {str(e)}")
                        if attempts >= max_attempts:
                            raise

                if result.get("outcome") == "HALT":
                    reason = result.get('reason_code', 'UNKNOWN_REASON')
                    raise RuntimeError(f"VEX Execution HALTED by Vanguard: {reason}")
                
                if result.get("outcome") == "ESCALATE":
                    raise TimeoutError(f"VEX Escalation timed out after {attempts} attempts. Capsule ID: {capsule_id}")

                # 4. Store Capability Token
                if "capability_token" in result:
                    self.current_token = result["capability_token"]

                return result
            except Exception as e:
                print(f"VEX Transport/Gate Error: {str(e)}")
                raise

    async def build_capsule(self, tool_name: str, parameters: Dict[str, Any], intent_context: Optional[str] = None) -> Dict[str, Any]:
        """
        Manually construct a signed Evidence Capsule without dispatching it.
        """
        intent = {
            "schema": "vex/intent/v3",
            "aid": self.aid,
            "request_sha256": self._hash_object(parameters),
            "commands": [tool_name, parameters],
            "confidence": 1.0,
            "capabilities": ["sdk_execution"]
        }
        if intent_context:
            intent["intent_context"] = intent_context

        authority = {
            "capsule_id": str(uuid.uuid4()),
            "outcome": "ALLOW",
            "reason_code": "SDK_GENERATED",
            "trace_root": "00" * 32,
            "nonce": int(time.time() * 1000),
            "prev_hash": "00" * 32, # Start of chain
            "binding_status": "UNBOUND",
            "supervision": {
                "branch_completeness": 1.0,
                "contradictions": 0,
                "confidence": 1.0
            }
        }

        identity = {
            "aid": "00" * 32,
            "identity_type": "software_sim",
            "pcrs": {"0": "00" * 32}
        }

        witness = {
            "chora_node_id": "local_witness",
            "receipt_hash": "00" * 32,
            "timestamp": int(time.time())
        }

        witness_hash = VEPBuilder.hash_segment(witness, inclusive=False)

        # --- Phase 2: HPKE Encryption (Python) ---
        gate_pk_b64 = None
        try:
            gate_pk_b64 = await self.fetch_public_key()
        except:
            pass

        if gate_pk_b64:
            # Custom HPKE Seal (X25519-HKDF-SHA256, AES-128-GCM)
            # 1. Recipient Public Key
            recipient_pk_bytes = base64.b64decode(gate_pk_b64)
            # If it's SPKI, extract raw bytes. If raw, use directly.
            try:
                recipient_pub = x25519.X25519PublicKey.from_public_bytes(recipient_pk_bytes)
            except:
                recipient_pub = serialization.load_der_public_key(recipient_pk_bytes)
                
            # 2. Ephemeral Keypair
            ephemeral_priv = x25519.X25519PrivateKey.generate()
            ephemeral_pub_bytes = ephemeral_priv.public_key().public_bytes(
                encoding=serialization.Encoding.Raw,
                format=serialization.PublicFormat.Raw
            )
            
            # 3. Shared Secret (DHKEM)
            shared_secret = ephemeral_priv.exchange(recipient_pub)
            
            # 4. KDF (HKDF-SHA256)
            # info = b"hpke-js" or similar context? VEX uses b"vex/intent/v3"
            info = b"vex/intent/v3"
            derived = HKDF(
                algorithm=hashes.SHA256(),
                length=12 + 16, # 12b nonce + 16b key (AES-128)
                salt=None,
                info=info
            ).derive(shared_secret)
            
            nonce = derived[:12]
            key = derived[12:]
            
            # 5. Encrypt (AES-128-GCM)
            import jcs
            plaintext = jcs.canonicalize(intent)
            aesgcm = AESGCM(key)
            ciphertext = aesgcm.encrypt(nonce, plaintext, None)
            
            # Intent Pillar commitment is hash of ciphertext
            intent_hash = hashlib.sha256(ciphertext).hexdigest()
            
            # Add HPKE metadata
            intent["hpke"] = {
                "enc": base64.b64encode(ephemeral_pub_bytes).decode(),
                "ciphertext": base64.b64encode(ciphertext).decode(),
                "schema": "vex/intent/v3/encrypted"
            }
        else:
            intent_hash = VEPBuilder.hash_segment(intent)

        authority_hash = VEPBuilder.hash_segment(authority)
        identity_hash = VEPBuilder.hash_segment(identity)

        capsule_root = VEPBuilder.calculate_capsule_root(
            intent_hash=intent_hash,
            authority_hash=authority_hash,
            identity_hash=identity_hash,
            witness_hash=witness_hash
        )

        # Signing (Real Ed25519 Hardware Seal)
        # Uses provn_sdk.ProvnSDK.sign_claim(claim, private_key_hex)
        sdk = provn_sdk.ProvnSDK()
        claim = {"data": capsule_root, "timestamp": int(time.time())}
        signed_claim = sdk.sign_claim(claim, self.identity_key)
        signature = bytes.fromhex(signed_claim["signature"])
        import base64

        return {
            "intent": intent,
            "authority": authority,
            "identity": identity,
            "witness": witness,
            "intent_hash": intent_hash,
            "authority_hash": authority_hash,
            "identity_hash": identity_hash,
            "witness_hash": witness_hash,
            "capsule_root": capsule_root,
            "crypto": {
                "algo": "ed25519",
                "signature_scope": "capsule_root",
                "signature_b64": base64.b64encode(signature).decode(),
                "signature_raw": signature # Keep raw for binary spec
            }
        }

    def to_binary(self, capsule: Dict[str, Any]) -> bytes:
        """
        Serializes the Evidence Capsule into the v0x03 Binary Wire format.
        """
        import struct
        import jcs

        # --- Header (76 Bytes) ---
        # magic (3b), version (1b), aid (32b), root (32b), nonce (8b)
        header = b"VEP"
        header += struct.pack("B", 0x03)
        header += bytes.fromhex(capsule["identity"]["aid"])
        header += bytes.fromhex(capsule["capsule_root"])
        header += struct.pack(">Q", capsule["authority"]["nonce"])

        # --- TLV Body ---
        def pack_tlv(tag: int, data: bytes) -> bytes:
            return struct.pack(">BI", tag, len(data)) + data

        body = b""
        body += pack_tlv(0x01, jcs.canonicalize(capsule["intent"]))
        body += pack_tlv(0x02, jcs.canonicalize(capsule["authority"]))
        body += pack_tlv(0x03, jcs.canonicalize(capsule["identity"]))
        body += pack_tlv(0x05, jcs.canonicalize(capsule["witness"]))
        
        # Raw signature from build_capsule (64 bytes)
        sig = capsule["crypto"]["signature_raw"]
        body += pack_tlv(0x06, sig)

        return header + body + header

    def _hash_object(self, obj: Any) -> str:
        import jcs
        import hashlib
        canonical_json = jcs.canonicalize(obj)
        return hashlib.sha256(canonical_json).hexdigest()

